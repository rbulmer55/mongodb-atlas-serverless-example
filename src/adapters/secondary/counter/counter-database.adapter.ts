import { Counter } from '@models/counter';
import { connect } from '@shared/databases-services/counter-service/connection';
import { counterService } from '@shared/databases-services/counter-service/counter/services';
import { v4 as uuid } from 'uuid';

export const incrementCounter = async () => {
	console.log('Secondary Adapter: Connecting to database');
	await connect();

	console.log('Secondary Adapter: Retrieving counter Id');
	const counter = await counterService.getLatestCounter();

	console.log('Secondary Adapter: Incrementing counter');
	const updatedCounter = await counterService.incrementCounter(
		counter ? counter.id : uuid()
	);
	console.log('Counter Updated', updatedCounter);
	return updatedCounter as Counter;
};
