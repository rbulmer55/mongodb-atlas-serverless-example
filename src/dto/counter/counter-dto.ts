export interface CounterDto {
  id: string;
  count: number;
}
