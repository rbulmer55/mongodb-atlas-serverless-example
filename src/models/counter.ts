export type Counter = {
  id: string;
  count: number;
  updatedAt?: string;
  createdAt?: string;
};
