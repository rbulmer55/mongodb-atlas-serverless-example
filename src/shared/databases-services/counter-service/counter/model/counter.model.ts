import { Schema, model, Document } from 'mongoose';
import { _id, idVirtual } from '../../helpers/schema-constants';

const options = {
	strict: true,
	timestamps: true,
	toJson: {
		virtuals: true,
	},
};

export interface CounterModel extends Document {
	_id: string;
	count: number;
}

const name = 'Counters';
const counterProperties = {
	_id,
	count: {
		type: Number,
		required: true,
		public: true,
		default: 1,
	},
};

const schema = new Schema(counterProperties, options);

schema.virtual('id').get(idVirtual);

export const Counter = model<CounterModel>(name, schema);
