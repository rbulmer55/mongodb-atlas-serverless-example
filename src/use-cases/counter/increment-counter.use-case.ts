import { incrementCounter } from '@adapters/secondary/counter/counter-database.adapter';

export const incrementCounterUseCase = async () => {
	console.log('Business Logic: Adding one to counter');
	return await incrementCounter();
};
