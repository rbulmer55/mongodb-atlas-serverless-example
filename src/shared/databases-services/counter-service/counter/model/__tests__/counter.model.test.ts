import {Counter} from '../counter.model';

let counterModel: any;

describe('Counter model', () => {
  beforeEach(() => {
    counterModel = {
      _id: '851e52e0-9444-4436-8f3d-6ffe47580458',
      count: 3,
    };
  });

  test('should default the properties', () => {
    const counter = new Counter();
    expect(counter).toMatchSnapshot();
  });

  test('should create when supplied properties', () => {
    const counter = new Counter(counterModel);
    expect(counter).toMatchSnapshot();
  });

  describe('validate', () => {
    test('should validate id', () => {
      counterModel._id = 'invalid-not-a-uuid';
      const counter = new Counter(counterModel);
      const error = counter.validateSync();
      expect(error?.errors).toMatchSnapshot();
    });

    test('should create id', () => {
      delete counterModel._id;
      const counter = new Counter(counterModel);
      expect(counter.id).not.toEqual(counterModel._id);
      expect(counter).toMatchSnapshot();
    });
  });
});
