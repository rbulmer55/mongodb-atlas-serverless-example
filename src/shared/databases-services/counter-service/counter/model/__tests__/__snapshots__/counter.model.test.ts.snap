// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter model should create when supplied properties 1`] = `
{
  "_id": "851e52e0-9444-4436-8f3d-6ffe47580458",
  "count": 3,
}
`;

exports[`Counter model should default the properties 1`] = `
{
  "_id": "1b9d6bcd-bbfd-4b2d-9b5d-ab8dfbbd4bed",
  "count": 1,
}
`;

exports[`Counter model validate should create id 1`] = `
{
  "_id": "1b9d6bcd-bbfd-4b2d-9b5d-ab8dfbbd4bed",
  "count": 3,
}
`;

exports[`Counter model validate should validate id 1`] = `
{
  "_id": [ValidatorError: Validator failed for path \`_id\` with value \`invalid-not-a-uuid\`],
}
`;
